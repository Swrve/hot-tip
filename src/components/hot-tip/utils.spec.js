import {
  getPosition,
  getBottom,
  getTop,
  getRight,
  getLeft
} from './utils';

describe('Utils', function() {
  const getBounds = (
    left = 0,
    top = 0,
    width = 80,
    height = 10,
    clientWidth = 800,
    clientHeight = 900) => {

    return {top, left, width, height, clientWidth, clientHeight}
  }

  describe('getPosition', function() {
    it('returns appropriate function (curries)', () => {
      expect(getPosition('top')).toEqual(getTop)
      expect(getPosition('right')).toEqual(getRight)
      expect(getPosition('left')).toEqual(getLeft)
      expect(getPosition('bottom')).toEqual(getBottom)
      expect(getPosition()).toEqual(getBottom)
    })
  })

  describe('getBottom/getTop horizontal offsets', function() {
    describe('getBottom', function() {
      it('Should not offset when mid is 100px from edge', () => {
        const position = getBottom(getBounds(60, 300))

        expect(position.layoutClass).toEqual('centred')
        expect(position.positionClass).toEqual('bottom')
        expect(position.x).toEqual(100)
        expect(position.y).toEqual(310)
        expect(position).toMatchSnapshot()
      })

      it('Should not offset when mid is 100px from (right) edge', () => {
        const position = getBottom(getBounds(660, 300))

        expect(position.layoutClass).toEqual('centred')
        expect(position.positionClass).toEqual('bottom')
        expect(position.x).toEqual(700)
        expect(position.y).toEqual(310)
        expect(position).toMatchSnapshot()
      })

      it('Should offset to left when < 100px from right', () => {
        const position = getBottom(getBounds(661, 300))

        expect(position.layoutClass).toEqual('right-offset')
        expect(position.positionClass).toEqual('bottom')
        expect(position.x).toEqual(701)
        expect(position.y).toEqual(310)
        expect(position).toMatchSnapshot()
      })

      it('Should offset to right when mid < 100px from left', () => {
        const position = getBottom(getBounds(59, 300))

        expect(position.layoutClass).toEqual('left-offset')
        expect(position.positionClass).toEqual('bottom')
        expect(position.x).toEqual(99)
        expect(position.y).toEqual(310)
        expect(position).toMatchSnapshot()
      })
    })

    describe('getTop', function() {
      it('Should not offset when mid is 100px from edge', () => {
        const position = getTop(getBounds(60, 300))

        expect(position.layoutClass).toEqual('centred')
        expect(position.positionClass).toEqual('top')
        expect(position.x).toEqual(100)
        expect(position.y).toEqual(300)
        expect(position).toMatchSnapshot()
      })

      it('Should not offset when mid is 100px from (right) edge', () => {
        const position = getTop(getBounds(660, 300))

        expect(position.layoutClass).toEqual('centred')
        expect(position.positionClass).toEqual('top')
        expect(position.x).toEqual(700)
        expect(position.y).toEqual(300)
        expect(position).toMatchSnapshot()
      })

      it('Should offset to left when < 100px from right', () => {
        const position = getTop(getBounds(661, 300))

        expect(position.layoutClass).toEqual('right-offset')
        expect(position.positionClass).toEqual('top')
        expect(position.x).toEqual(701)
        expect(position.y).toEqual(300)
        expect(position).toMatchSnapshot()
      })

      it('Should offset to right when mid < 100px from left', () => {
        const position = getTop(getBounds(59, 300))

        expect(position.layoutClass).toEqual('left-offset')
        expect(position.positionClass).toEqual('top')
        expect(position.x).toEqual(99)
        expect(position.y).toEqual(300)
        expect(position).toMatchSnapshot()
      })
    })
  })

  describe('getLeft/getRight vertical offsets', function() {
    describe('getRight', function() {
      it('Should not offset when vertical mid is 40px from top', () => {
        const position = getRight(getBounds(400, 45))

        expect(position.layoutClass).toEqual('mid')
        expect(position.positionClass).toEqual('right')
        expect(position.x).toEqual(480)
        expect(position.y).toEqual(50)
        expect(position).toMatchSnapshot()
      })

      it('Should not offset when vertical mid is 40px from bottom', () => {
        const position = getRight(getBounds(400, 845))

        expect(position.layoutClass).toEqual('mid')
        expect(position.positionClass).toEqual('right')
        expect(position.x).toEqual(480)
        expect(position.y).toEqual(850)
        expect(position).toMatchSnapshot()
      })

      it('Should offset to left when < 100px from right', () => {
        const position = getRight(getBounds(400, 44))

        expect(position.layoutClass).toEqual('top-offset')
        expect(position.positionClass).toEqual('right')
        expect(position.x).toEqual(480)
        expect(position.y).toEqual(49)
        expect(position).toMatchSnapshot()
      })

      it('Should offset to right when mid < 100px from left', () => {
        const position = getRight(getBounds(400, 846))

        expect(position.layoutClass).toEqual('bottom-offset')
        expect(position.positionClass).toEqual('right')
        expect(position.x).toEqual(480)
        expect(position.y).toEqual(851)
        expect(position).toMatchSnapshot()
      })
    })

    describe('getLeft', function() {
      it('Should not offset when vertical mid is 40px from top', () => {
        const position = getLeft(getBounds(400, 45))

        expect(position.layoutClass).toEqual('mid')
        expect(position.positionClass).toEqual('left')
        expect(position.x).toEqual(400)
        expect(position.y).toEqual(50)
        expect(position).toMatchSnapshot()
      })

      it('Should not offset when vertical mid is 40px from bottom', () => {
        const position = getLeft(getBounds(400, 845))

        expect(position.layoutClass).toEqual('mid')
        expect(position.positionClass).toEqual('left')
        expect(position.x).toEqual(400)
        expect(position.y).toEqual(850)
        expect(position).toMatchSnapshot()
      })

      it('Should offset to left when < 100px from right', () => {
        const position = getLeft(getBounds(400, 44))

        expect(position.layoutClass).toEqual('top-offset')
        expect(position.positionClass).toEqual('left')
        expect(position.x).toEqual(400)
        expect(position.y).toEqual(49)
        expect(position).toMatchSnapshot()
      })

      it('Should offset to right when mid < 100px from left', () => {
        const position = getLeft(getBounds(400, 846))

        expect(position.layoutClass).toEqual('bottom-offset')
        expect(position.positionClass).toEqual('left')
        expect(position.x).toEqual(400)
        expect(position.y).toEqual(851)
        expect(position).toMatchSnapshot()
      })
    })
  })

  describe('getTop/bottom/left/right swap to opposite if too close to bounds', () => {
    it('getBottom returns top position if too close to bottom', () => {
      const bottom = getBottom(getBounds(360, 849))
      expect(bottom.positionClass).toEqual('bottom')
      expect(bottom.y).toEqual(859)
      const top = getBottom(getBounds(360, 850))
      expect(top.positionClass).toEqual('top')
      expect(top.y).toEqual(850)
      expect({bottom, top}).toMatchSnapshot()
    })

    it('getTop returns bottom position if too close to top', () => {
      const bottom = getTop(getBounds(360, 40))
      expect(bottom.positionClass).toEqual('bottom')
      expect(bottom.y).toEqual(50)
      const top = getTop(getBounds(360, 41))
      expect(top.positionClass).toEqual('top')
      expect(top.y).toEqual(41)
      expect({bottom, top}).toMatchSnapshot()
    })

    it('getLeft returns right position if too close to left', () => {
      const left = getLeft(getBounds(201, 445))
      expect(left.positionClass).toEqual('left')
      expect(left.x).toEqual(201)
      const right = getLeft(getBounds(200, 445))
      expect(right.positionClass).toEqual('right')
      expect(right.x).toEqual(280)
      expect({right, left}).toMatchSnapshot()
    })

    it('getRight returns left position if too close to right', () => {
      const right = getRight(getBounds(519, 445))
      expect(right.positionClass).toEqual('right')
      expect(right.x).toEqual(599)
      const left = getRight(getBounds(520, 445))
      expect(left.positionClass).toEqual('left')
      expect(left.x).toEqual(520)
      expect({right, left}).toMatchSnapshot()
    })
  })
})
