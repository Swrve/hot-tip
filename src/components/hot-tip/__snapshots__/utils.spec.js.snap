// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils getBottom/getTop horizontal offsets getBottom Should not offset when mid is 100px from (right) edge 1`] = `
Object {
  "layoutClass": "centred",
  "positionClass": "bottom",
  "x": 700,
  "y": 310,
}
`;

exports[`Utils getBottom/getTop horizontal offsets getBottom Should not offset when mid is 100px from edge 1`] = `
Object {
  "layoutClass": "centred",
  "positionClass": "bottom",
  "x": 100,
  "y": 310,
}
`;

exports[`Utils getBottom/getTop horizontal offsets getBottom Should offset to left when < 100px from right 1`] = `
Object {
  "layoutClass": "right-offset",
  "positionClass": "bottom",
  "x": 701,
  "y": 310,
}
`;

exports[`Utils getBottom/getTop horizontal offsets getBottom Should offset to right when mid < 100px from left 1`] = `
Object {
  "layoutClass": "left-offset",
  "positionClass": "bottom",
  "x": 99,
  "y": 310,
}
`;

exports[`Utils getBottom/getTop horizontal offsets getTop Should not offset when mid is 100px from (right) edge 1`] = `
Object {
  "layoutClass": "centred",
  "positionClass": "top",
  "x": 700,
  "y": 300,
}
`;

exports[`Utils getBottom/getTop horizontal offsets getTop Should not offset when mid is 100px from edge 1`] = `
Object {
  "layoutClass": "centred",
  "positionClass": "top",
  "x": 100,
  "y": 300,
}
`;

exports[`Utils getBottom/getTop horizontal offsets getTop Should offset to left when < 100px from right 1`] = `
Object {
  "layoutClass": "right-offset",
  "positionClass": "top",
  "x": 701,
  "y": 300,
}
`;

exports[`Utils getBottom/getTop horizontal offsets getTop Should offset to right when mid < 100px from left 1`] = `
Object {
  "layoutClass": "left-offset",
  "positionClass": "top",
  "x": 99,
  "y": 300,
}
`;

exports[`Utils getLeft/getRight vertical offsets getLeft Should not offset when vertical mid is 40px from bottom 1`] = `
Object {
  "layoutClass": "mid",
  "positionClass": "left",
  "x": 400,
  "y": 850,
}
`;

exports[`Utils getLeft/getRight vertical offsets getLeft Should not offset when vertical mid is 40px from top 1`] = `
Object {
  "layoutClass": "mid",
  "positionClass": "left",
  "x": 400,
  "y": 50,
}
`;

exports[`Utils getLeft/getRight vertical offsets getLeft Should offset to left when < 100px from right 1`] = `
Object {
  "layoutClass": "top-offset",
  "positionClass": "left",
  "x": 400,
  "y": 49,
}
`;

exports[`Utils getLeft/getRight vertical offsets getLeft Should offset to right when mid < 100px from left 1`] = `
Object {
  "layoutClass": "bottom-offset",
  "positionClass": "left",
  "x": 400,
  "y": 851,
}
`;

exports[`Utils getLeft/getRight vertical offsets getRight Should not offset when vertical mid is 40px from bottom 1`] = `
Object {
  "layoutClass": "mid",
  "positionClass": "right",
  "x": 480,
  "y": 850,
}
`;

exports[`Utils getLeft/getRight vertical offsets getRight Should not offset when vertical mid is 40px from top 1`] = `
Object {
  "layoutClass": "mid",
  "positionClass": "right",
  "x": 480,
  "y": 50,
}
`;

exports[`Utils getLeft/getRight vertical offsets getRight Should offset to left when < 100px from right 1`] = `
Object {
  "layoutClass": "top-offset",
  "positionClass": "right",
  "x": 480,
  "y": 49,
}
`;

exports[`Utils getLeft/getRight vertical offsets getRight Should offset to right when mid < 100px from left 1`] = `
Object {
  "layoutClass": "bottom-offset",
  "positionClass": "right",
  "x": 480,
  "y": 851,
}
`;

exports[`Utils getTop/bottom/left/right swap to opposite if too close to bounds getBottom returns top position if too close to bottom 1`] = `
Object {
  "bottom": Object {
    "layoutClass": "centred",
    "positionClass": "bottom",
    "x": 400,
    "y": 859,
  },
  "top": Object {
    "layoutClass": "centred",
    "positionClass": "top",
    "x": 400,
    "y": 850,
  },
}
`;

exports[`Utils getTop/bottom/left/right swap to opposite if too close to bounds getLeft returns right position if too close to left 1`] = `
Object {
  "left": Object {
    "layoutClass": "mid",
    "positionClass": "left",
    "x": 201,
    "y": 450,
  },
  "right": Object {
    "layoutClass": "mid",
    "positionClass": "right",
    "x": 280,
    "y": 450,
  },
}
`;

exports[`Utils getTop/bottom/left/right swap to opposite if too close to bounds getRight returns left position if too close to right 1`] = `
Object {
  "left": Object {
    "layoutClass": "mid",
    "positionClass": "left",
    "x": 520,
    "y": 450,
  },
  "right": Object {
    "layoutClass": "mid",
    "positionClass": "right",
    "x": 599,
    "y": 450,
  },
}
`;

exports[`Utils getTop/bottom/left/right swap to opposite if too close to bounds getTop returns bottom position if too close to top 1`] = `
Object {
  "bottom": Object {
    "layoutClass": "centred",
    "positionClass": "bottom",
    "x": 400,
    "y": 50,
  },
  "top": Object {
    "layoutClass": "centred",
    "positionClass": "top",
    "x": 400,
    "y": 41,
  },
}
`;
